walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
+      HashSet<String> meetingAttendees = new HashSet<String>(attendees);
+      eventAttendees.retainAll(meetingAttendees); // set intersection
+      if (eventAttendees.size() > 0) {
+        eventsByStart.add(event);
+      }
+    }
+    Collections.sort(eventsByStart, Event.ORDER_BY_START_TIME);   
+
+    // Check again to see if there are any events
+    if (eventsByStart.size() == 0 ) {
+      possibleTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()));
+      return possibleTimes;
+    }
+
+    // Check if we can have a meeting before the earliest-starting event
+    int earliestStart = eventsByStart.get(0).getWhen().start();
+    if (TimeRange.START_OF_DAY < earliestStart && earliestStart - TimeRange.START_OF_DAY >= duration) {
+      possibleTimes.add(TimeRange.fromStartEnd(
+          TimeRange.START_OF_DAY, earliestStart, false));
+    }
+
+    // Check in between events - we know we can't have anything until
+    // the earliest-starting event ends, so find the earliest-starting event
+    // after that end time and see if the meeting fits between them
+    Event currentEarliestEvent = eventsByStart.get(0);
+    Event nextEvent;
+    while (true) {
+      // Try to get event
+      Optional<Event> potentialNextEvent = getNextEarliestEvent(
+          currentEarliestEvent.getWhen().end(), eventsByStart, 0, eventsByStart.size()-1);
+
+      if (!potentialNextEvent.isPresent()) { 
+        break; // No event starts after current one ends
+      }
+
+      // Check if we can fit the meeting between the two events
+      // ***BUG***
+      nextEvent = potentialNextEvent.get();
+      int nextEventStart = nextEvent.getWhen().start();
+      int currentEventEnd = currentEarliestEvent.getWhen().end();
+      if (nextEventStart - currentEventEnd >= duration) {
+        possibleTimes.add(TimeRange.fromStartEnd(
+            currentEventEnd, nextEventStart, false));
+      }
+      currentEarliestEvent = nextEvent;
+    }
+
+
+    // Lastly, check if we can have a meeting before the latest-ending event
+    ArrayList<Event> eventsByEnd = new ArrayList<Event>(eventsByStart);
+    Collections.sort(eventsByEnd, Event.ORDER_BY_END_TIME);
+    int latestEventEnd = eventsByEnd.get(eventsByEnd.size()-1).getWhen().end();
+    if (TimeRange.END_OF_DAY - latestEventEnd >= duration) {
+      possibleTimes.add(TimeRange.fromStartEnd(
+        latestEventEnd, TimeRange.END_OF_DAY, true));
+    }
+
+    return possibleTimes;
+  }
+
+  // Parameters: int endTime, representing a time in minutes 
+  //     List<Event>, a list of events which must be already sorted by start time
+  //     ints low and high, marking off the section of list of events to be searched
+  // Returns: the earliest-starting event that starts at or after endTime,
+  // or an empty optional if there is no such event listed
+  private Optional<Event> getNextEarliestEvent(int endTime, 
+      List<Event> events, int low, int high) {
```

Comment:
It seems you always start with all events, so low and high don't need to be parameters

Reply:
I use them for the recursion inside the function.